Net Runner/Settings/AddModelTableViewController.mm

```diff
@@ -0,0 +1,293 @@
+    
+    self.downloadLabel.hidden = YES;
+    self.downloadProgressView.hidden = YES;
+}
+
+// MARK: - Table View Delegate
+
+- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
+    return UITableViewAutomaticDimension;
+}
+
+// MARK: - Text Field Delegate
+
+- (BOOL)textFieldShouldReturn:(UITextField *)textField {
+    [textField resignFirstResponder];
+    return YES;
+}
+
+// MARK: - User Interaction
+
+- (IBAction)cancelAddModel:(id)sender {
+    [self resignFirstResponder];
+    [self dismissViewControllerAnimated:YES completion:nil];
+}
+
+- (IBAction)visitNetRunnerRepository:(id)sender {
+    [UIApplication.sharedApplication openURL:[NSURL URLWithString:NetRunnerGitHubRepository]];
+}
+
+- (IBAction)importModel:(id)sender {
+    
+    if ( self.URLField.text.length == 0 ) {
+        NSLog(@"Enter a URL");
+        [self showError:NetRunnerAddModelInvalidURLError()];
+        return;
+    }
+    
+    NSURL *URL = [NSURL URLWithString:self.URLField.text];
+    
+    if ( URL == nil ) {
+        NSLog(@"Invalid URL");
+        [self showError:NetRunnerAddModelInvalidURLError()];
+        return;
+    }
+    
+    if ( self.importer != nil ) {
+        [self.importer cancel];
+        self.importer = nil;
+    }
+    
+    self.importer = [[ModelImporter alloc] initWithURL:URL delegate:self destinationDirectory:[NSURL fileURLWithPath:[self modelsPath]]];
+    [self.importer import];
+}
+
+// MARK: - Model Importer Delegate
+
+- (BOOL (^_Nullable)(NSString *path, NSDictionary *JSON, NSError **error))modelImporter:(ModelImporter*)importer validationBlockForModelBundleAtURL:(NSURL*)URL {
+    
+    // Net Runner requires a single input that is of type "image",
+    // whose "shape" is 3 elements with the last element being 3 for three color channels
+    
+    return ^BOOL(NSString *path, NSDictionary *JSON, NSError **error) {
+        
+        NSArray *inputs = JSON[@"inputs"];
+        
+        if ( inputs.count != 1 ) {
+            *error = NetRunnerModelInputsError();
+            return NO;
+        }
+        
+        NSDictionary *input = inputs[0];
+        
+        if ( ![input[@"type"] isEqualToString:@"image"] ) {
+            *error = NetRunnerModelInputsError();
+            return NO;
+        }
+        
+        if ( [input[@"shape"] count] != 3 || [input[@"shape"][2] integerValue] != 3 ) {
+            *error = NetRunnerImageShapeError();
+            return NO;
+        }
+        
+        return YES;
+    };
+}
+
+- (void)modelImporterDownloadDidBegin:(ModelImporter*)importer {
+    NSLog(@"Importer Did Begin Download");
+    
+    dispatch_async(dispatch_get_main_queue(), ^{
+        self.downloadProgressView.hidden = NO;
+        self.downloadProgressView.progress = 0;
+        self.downloadLabel.hidden = NO;
+    });
+}
+
+- (void)modelImporter:(ModelImporter*)importer downloadDidProgress:(float)progress {
+    NSLog(@"Importer Did Progress");
```

Comment:
Getting failures with the following message:
```
2018-09-25 15:21:35.426042-0700 Net Runner[58280:1198866] Task <0A2FBCEA-D8BD-4811-9A24-1D6001AB31E9>.<1> finished with error - code: -1002
2018-09-25 15:23:51.219895-0700 Net Runner[58280:1191531] Importer Did Begin Download
2018-09-25 15:23:51.220704-0700 Net Runner[58280:1199591] Task <5F87FC8D-70F2-43FF-BB19-F6BEBF50463C>.<1> finished with error - code: -1002
2018-09-25 15:26:35.202885-0700 Net Runner[58280:1191531] Importer Did Begin Download
2018-09-25 15:26:35.203130-0700 Net Runner[58280:1203627] Task <1DAE37BE-A89F-4439-8F62-56ECB5E95F66>.<1> finished with error - code: -1002
```

Reply:
The issue with those seems to have been that I was exposing the zipfile using Python's `SimpleHTTPServer` over `localhost` and the simulated iPhone does not understand how to look up on `localhost`?

Putting it on tinyupload led to an error saying that the file could not be unzipped.

I created the zipfile as follows:
1. I took the phenomenal-face model, renamed it to Dark face, and kept everything else as is.
2. I renamed the tfbundle directory to dark-face.tfbundle
3. I ran `zip -r dark-face.tfbundle.zip dark-face.tfbundle/*`

This seems not to work. I will try again using the Finder compress option. Please add a command line equivalent, though.

Reply:
Doing this with a finder-compressed file led to the same error that I was seeing earlier:
```
2018-09-25 15:38:07.806290-0700 Net Runner[58280:1191531] Importer Did Begin Download
2018-09-25 15:38:07.806718-0700 Net Runner[58280:1211738] Task <EFD8FF0B-9660-43CA-BE30-8F360386D5E3>.<1> finished with error - code: -1002
```

This suggests that I input the URL incorrectly. The error should surface in the UI.

Reply:
The upload is not working with the model uploaded here: https://files.fm/u/9d936z3k

@phildow : You can use that to debut, I may have made some rookie mistake.

We should not merge until this experience is better.

Reply:
@nkashy1 I was always testing with SimpleHTTPServer on localhost and it worked fine on my end in the simulator. I zipped the .tfbundle folder itself. And errors should definitely surface to the UI, not sure why this one isn't. Let me try with that file on my end.

Reply:
Alright, there were a lot of mistakes on my side:
1. I hadn't `pod install`ed the latest TensorIO or the zip library
2. I missed the `http://` prefix on the model URL
3. The files.fm link downloaded an HTML page that provided a download button for the file (so was not a good test).

At the end of the day, running `SimpleHTTPServer` worked, for a zipfile made by the Finder compress as well as a zipfile made with the `zip` command line utility.

The only thing to change, I would say, is validate that there is an `http://` or `https://` prefix in the tfbundle zip URL.